### PR TITLE
2.4 xb 1548

### PR DESCRIPTION
**Pull request by altmannmarcelo**
This implements #PXB-1548.

Buffer pool dump happens at the beginning of backup if `--dump-innodb-buffer-pool` is set. 
Prior to copying `ib_buffer_pool` it checks if current `SHOW STATUS LIKE 'Innodb_buffer_pool_dump_status'` contains the string `dump completed at`. If dump is still in progress, the tool keeps checking untill `--dump-innodb-buffer-pool-timeout=N` seconds has elapsed since we originally started the dump.

User can choose to change the default `innodb_buffer_pool_dump_pct`. If `--dump-innodb-buffer-pool-pct` is set, it stores the current MySQL `innodb_buffer_pool_dump_pct` value, then it changes it to the desired percentage. After the check in the end of backup, orginal values is restored back.

3 new parameters have been added:
`--dump-innodb-buffer-pool` - When set, this will trigger dump of buffer pool prior to copying ib_buffer_pool
`--dump-innodb-buffer-pool-timeout=N` - This controlls for how long (in seconds) to wait for dump of buffer pool before moving on. (default: 10)
`--dump-innodb-buffer-pool-pct=N` - This controlls the percentage of buffer pool to dump. If set, xtrabackup will save current MySQL value, change to the desired value, once dump is completed or it aborts due to time out, it will change it back to original value.

### CODE REVIEW

**Comment by gl-sergei** [comment]
Hi @altmannmarcelo ! Are you willing to continue to work on this patch? If so, I would ask to you to start with adding a test case (you can see examples in xtrabackup/test/t/ directory). Then fix code formatting. Please also make sure to check that patch doesn't break anything on older MySQL versions.

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 1]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+		has_innodb_buffer_pool_dump_pct()) {
+		mysql_variable variables[] = {
+			{"innodb_buffer_pool_dump_pct", &innodb_buffer_pool_dump_pct},
+			{NULL, NULL}
+		};
+		read_mysql_variables(connection,
+			"SHOW GLOBAL VARIABLES LIKE 'innodb_buffer_pool_dump_pct'", 
+			variables, true);
+
+		original_innodb_buffer_pool_dump_pct =
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: @gl-sergei I'm not sure is ok to break the value assignment into 2 line, but leaving it on a single line would exceed 80 characters.

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 2]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+		mysql_variable variables[] = {
+			{"innodb_buffer_pool_dump_pct", &innodb_buffer_pool_dump_pct},
+			{NULL, NULL}
+		};
+		read_mysql_variables(connection,
+			"SHOW GLOBAL VARIABLES LIKE 'innodb_buffer_pool_dump_pct'", 
+			variables, true);
+
+		original_innodb_buffer_pool_dump_pct = 
+		atoi(innodb_buffer_pool_dump_pct);
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: @gl-sergei I'm not sure is ok to break the value assignment into 2 line, but leaving it on a single line would exceed 80 characters.

**Comment by altmannmarcelo** [comment]
Tested on not supported version and it works fine. Even if you pass `--dump-innodb-buffer-pool` it won't attempt to run the dump.

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 3]
```diff
@@ -1559,6 +1559,11 @@ backup_finish()
 
 	/* Copy buffer pool dump or LRU dump */
 	if (!opt_rsync) {
+		if(opt_dump_innodb_buffer_pool && has_innodb_buffer_pool_dump()) {
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: s/if(opt/if (opt/

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 4]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
 	mutex_free(&mdl_lock_con_mutex);
 }
 
+bool
+has_innodb_buffer_pool_dump()
+{
+	if((server_flavor == FLAVOR_PERCONA_SERVER ||
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: s/if((server_flavor/if ((server_flavor/

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 5]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
 
+bool
+has_innodb_buffer_pool_dump()
+{
+	if((server_flavor == FLAVOR_PERCONA_SERVER || 
+		server_flavor == FLAVOR_MYSQL) && mysql_server_version >= 50603) {
+		return true;
+	}
+		
+	if(server_flavor == FLAVOR_MARIADB && mysql_server_version >= 10000) {
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: s/if(server_flavor/if (server_flavor/

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 6]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+		return true;
+	}
+
+	return false;
+}
+
+bool
+has_innodb_buffer_pool_dump_pct()
+{
+	if((server_flavor == FLAVOR_PERCONA_SERVER ||
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: s/if((server_flavor/if ((server_flavor/

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 7]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+
+bool
+has_innodb_buffer_pool_dump_pct()
+{
+	if((server_flavor == FLAVOR_PERCONA_SERVER || 
+		server_flavor == FLAVOR_MYSQL) && mysql_server_version >= 50702) {
+		return true;
+	}
+		
+	if(server_flavor == FLAVOR_MARIADB && mysql_server_version >= 10110) {
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: s/if(server_flavor/if (server_flavor/

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 8]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+void
+dump_innodb_buffer_pool(MYSQL *connection)
+{
+	innodb_buffer_pool_dump_start_time = (ssize_t)my_time(MY_WME);
+
+	char *innodb_buffer_pool_dump_pct;
+	char *change_bp_dump_pct_query;
+
+	/* Verify if we need to change innodb_buffer_pool_dump_pct */
+	if(opt_dump_innodb_buffer_pool_pct != 0 &&
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: s/if(opt_dump_innodb_buffer_pool_pct/if (opt_dump_innodb_buffer_pool_pct/

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 9]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+	innodb_buffer_pool_dump_start_time = (ssize_t)my_time(MY_WME);
+
+	char *innodb_buffer_pool_dump_pct;
+	char *change_bp_dump_pct_query;
+
+	/* Verify if we need to change innodb_buffer_pool_dump_pct */
+	if(opt_dump_innodb_buffer_pool_pct != 0 && 
+		has_innodb_buffer_pool_dump_pct()) {
+		mysql_variable variables[] = {
+			{"innodb_buffer_pool_dump_pct", &innodb_buffer_pool_dump_pct},
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: Line is too long

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 10]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+			{NULL, NULL}
+		};
+		read_mysql_variables(connection,
+			"SHOW GLOBAL VARIABLES LIKE 'innodb_buffer_pool_dump_pct'",
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: Line is too long and extra trailing space

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 2 above)
Comment (gl-sergei) [reply storage/innobase/xtrabackup/src/backup_mysql.cc]: It is fine to break, please also add extra tab on second line

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 11]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+			variables, true);
+
+		original_innodb_buffer_pool_dump_pct = 
+		atoi(innodb_buffer_pool_dump_pct);
+
+		free_mysql_variables(variables);
+		xb_a(asprintf(&change_bp_dump_pct_query,
+				"SET GLOBAL innodb_buffer_pool_dump_pct = %u",
+				opt_dump_innodb_buffer_pool_pct));
+		xb_mysql_query(mysql_connection, change_bp_dump_pct_query, false);
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: This line is too long

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 12]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+check_dump_innodb_buffer_pool(MYSQL *connection)
+{
+	const ssize_t timeout = opt_dump_innodb_buffer_pool_timeout;
+
+	char *innodb_buffer_pool_dump_status;
+	char *change_bp_dump_pct_query;
+
+
+	mysql_variable status[] = {
+			{"Innodb_buffer_pool_dump_status", &innodb_buffer_pool_dump_status},
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: This line and entire next block are too long

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 13]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+		read_mysql_variables(connection,
+			"SHOW STATUS LIKE 'Innodb_buffer_pool_dump_status'", status, true);
+
+		os_thread_sleep(1000000);
+	}
+
+	free_mysql_variables(status);
+
+	/* restore original innodb_buffer_pool_dump_pct */
+	if(opt_dump_innodb_buffer_pool_pct != 0 &&
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: s/if(opt_dump_innodb_buffer_pool/if (opt_dump_innodb_buffer_pool/

File: storage/innobase/xtrabackup/src/xtrabackup.cc [hunk 14]
```diff
@@ -4712,6 +4737,12 @@ xtrabackup_backup_func(void)
 		exit(EXIT_FAILURE);
 	}
 
+	if (!opt_rsync) {
+		if(opt_dump_innodb_buffer_pool && has_innodb_buffer_pool_dump()) {
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/xtrabackup.cc]: s/if(opt_dump_innodb_buffer_pool/if (opt_dump_innodb_buffer_pool/

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 15]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+}
+
+bool
+has_innodb_buffer_pool_dump_pct()
+{
+	if((server_flavor == FLAVOR_PERCONA_SERVER || 
+		server_flavor == FLAVOR_MYSQL) && mysql_server_version >= 50702) {
+		return true;
+	}
+
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc:2193]: Whitespace on empty line.

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 16]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
 }
 
+bool
+has_innodb_buffer_pool_dump()
+{
+	if((server_flavor == FLAVOR_PERCONA_SERVER || 
+		server_flavor == FLAVOR_MYSQL) && mysql_server_version >= 50603) {
+		return true;
+	}
+
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc:2177]: Whitespace on empty line

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 4 above)
Comment (gl-sergei) [reply storage/innobase/xtrabackup/src/backup_mysql.cc]: trailing whitespace

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 6 above)
Comment (gl-sergei) [reply storage/innobase/xtrabackup/src/backup_mysql.cc]: trailing whitespace

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 8 above)
Comment (gl-sergei) [reply storage/innobase/xtrabackup/src/backup_mysql.cc]: trailing whitespace

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 10 above)
Comment (gl-sergei) [reply storage/innobase/xtrabackup/src/backup_mysql.cc]: trailing whitespace

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 1 above)
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: trailing whitespace

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 17]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+			{NULL, NULL}
+	};
+
+	read_mysql_variables(connection,
+		"SHOW STATUS LIKE 'Innodb_buffer_pool_dump_status'", status, true);
+
+	/* check if dump has been completed */
+	msg_ts("Checking if InnoDB buffer pool dump has completed\n");
+	while (!strstr(innodb_buffer_pool_dump_status, "dump completed at")) {
+		if(innodb_buffer_pool_dump_start_time +
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc]: Missing space after "if" and extra trailing whitespace

File: storage/innobase/xtrabackup/src/backup_mysql.cc (hunk 13 above)
Comment (gl-sergei) [reply storage/innobase/xtrabackup/src/backup_mysql.cc]: trailing whitespace

File: storage/innobase/xtrabackup/src/backup_mysql.cc [hunk 18]
```diff
@@ -2160,3 +2165,113 @@ mdl_unlock_all()
...
+	}
+		
+	if(server_flavor == FLAVOR_MARIADB && mysql_server_version >= 10000) {
+		return true;
+	}
+
+	return false;
+}
+
+bool
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.cc:2186]: Please make it static

File: storage/innobase/xtrabackup/src/backup_mysql.h [hunk 19]
```diff
@@ -113,4 +113,16 @@ mdl_lock_table(ulint space_id);
 void
 mdl_unlock_all();
 
+bool
+has_innodb_buffer_pool_dump();
+
+bool
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.h:119]: Please make it static and local to `backup_mysql`

File: storage/innobase/xtrabackup/src/backup_mysql.h [hunk 20]
```diff
@@ -113,4 +113,16 @@ mdl_lock_table(ulint space_id);
 void
 mdl_unlock_all();
 
+bool
+has_innodb_buffer_pool_dump();
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_mysql.h:117]: Consider making it static and local to `backup_mysql`, `check_dump_buffer_pool` could be made the only entry point.

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 21]
```diff
@@ -1559,6 +1559,11 @@ backup_finish()
 
 	/* Copy buffer pool dump or LRU dump */
 	if (!opt_rsync) {
```
Comment (gl-sergei) [inline storage/innobase/xtrabackup/src/backup_copy.cc:1567]: Can we make it work with rsync?

**Review by gl-sergei — CHANGES_REQUESTED** [review]

**Comment by altmannmarcelo** [comment]
@gl-sergei adjusted code as per your review

**Comment by altmannmarcelo** [comment]
Implemented in a single commit over https://github.com/percona/percona-xtrabackup/pull/563 . Thanks, @gl-sergei for the review.

**Pull request closed without merging**